Metal Trace Changes: #8062 Prefetcher stall_flag and Dispatcher wait clear_count sem change

kmabeeTT opened the pull request:
Hi @pgkeller here are 2 changes discussed a week ago for the trace hang I was facing (now resolved via https://github.com/tenstorrent/tt-metal/issues/7793, this wasn't needed) that we said would be good to submit.  

I wil probably squash the 3 related prefetcher-stall changes if you like them. 

I went back and forth with the implementation a bit (whether to early return in `read_from_pcie()` and rely on code in caller function to barrier and increase fence on the cmd being stalled on,  or early return in `fetch_q_get_cmds()` but add a few lines to end of `read_from_pcie()` to handle the barrier of the cmd being stalled on, so I left them as seperate commits, in the end I think the version I left it off with is safest (early exit at the top of `fetch_q_get_cmds()`)

Don't have any tests that seem to expose bugs / rely on these changes though.

kmabeeTT committed:
#8062: Add Prefetcher Stall flag used by CQ_PREFETCH_CMD_EXEC_BUF (FetchQ MSB)

 - Host sets MSB of FetchQ entry on ExecBuf cmd to denote
   that prefetcher should stall and not fetch any more cmds
   since ExecBuf will read TraceBuffer data and write to CmdDataQ
   which will clobber subsequently fetched cmds. Remove previous
   "ugly hack" that was doing similiar thing.

 - On STALL_NEXT, barrier/wait for fetched cmd requesting a stall to
   return, and increase fence, before moving to STALLED state and
   early exit in fetch_q_get_cmds() when STALLED

 - PR Feedback and fix for NOT_STALLED setting
 - Add assert to make sure ExecBuf is comes with stall_flag=true, it's
   required now that ugly-hack is removed, otherwise hang.
 - Update test_prefetcher.cpp to set stall flag (FetchQ MSB) for
   ExecBuf, otherwise hang. Needed to change to make cmd_sizes uint32_t
   instead of uint16_t through the code since currently today FetchQ entry size
   dispatch_constants::prefetch_q_entry_type is uint32_t, to be able to carry
   though the MSB bit properly.

kmabeeTT committed:
#0: Change CQ_DISPATCH_CMD_WAIT(clear_count=1) impl in cq_dispatcher.cpp

 - Use noc_semaphore_inc(neg_val) + noc_async_atomic_barrier() as per
   Paul feedback, instead of ~unsafe direct set to ptr value

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -149,6 +157,10 @@ void read_from_pcie(volatile tt_l1_ptr uint32_t *& prefetch_q_rd_ptr,
     }
+
+    // NOT_STALLED Persists. STALL_NEXT -> STALLED.
+    stall_state = static_cast<StallState>(stall_state << 1);
```
given this relies on specific values of the enum, let's assign the enum values explicitly, eg, NOT_STALLED=0

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -184,7 +197,10 @@ void fetch_q_get_cmds(uint32_t& fence, uint32_t& cmd_ptr, uint32_t& pcie_read_pt
-    uint32_t fetch_size = (uint32_t)*prefetch_q_rd_ptr << prefetch_q_log_minsize;
+
+    uint32_t fetch_size = ((uint32_t)(*prefetch_q_rd_ptr) & ~prefetch_q_msb_mask) << prefetch_q_log_minsize;
+    bool stall_flag = (*prefetch_q_rd_ptr & prefetch_q_msb_mask) != 0;
```
prefetch_q_rd_ptr is a volatile L1 address, only read it once (and store result in a local register)

kmabeeTT replied:
Good point..thanks!  This seems like the right pattern (applied to both places in code).

```
    uint32_t prefetch_q_rd_ptr_local = *prefetch_q_rd_ptr;
    uint32_t fetch_size = (prefetch_q_rd_ptr_local & ~prefetch_q_msb_mask) << prefetch_q_log_minsize;
    bool stall_flag = (prefetch_q_rd_ptr_local & prefetch_q_msb_mask) != 0;```

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -117,6 +120,11 @@ void read_from_pcie(volatile tt_l1_ptr uint32_t *& prefetch_q_rd_ptr,
                     uint32_t size) {
 
+    // When stalled, prevent prefetcher from fetching next cmd via early exit.
+    if (stall_state == STALLED) {
```
I'm thinking we want this test in get_cmds so that we don't even read the fetch_q when stalled

kmabeeTT replied:
Okay cool yes I agree, this version you commented on was outdated, and the version after the followup commit has it in fetch_q_get_cmds(), will keep it there.

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -184,7 +197,10 @@ void fetch_q_get_cmds(uint32_t& fence, uint32_t& cmd_ptr, uint32_t& pcie_read_pt
+
+    uint32_t fetch_size = ((uint32_t)(*prefetch_q_rd_ptr) & ~prefetch_q_msb_mask) << prefetch_q_log_minsize;
+    bool stall_flag = (*prefetch_q_rd_ptr & prefetch_q_msb_mask) != 0;
+    stall_state = stall_flag ? STALL_NEXT : stall_state; // NOT_STALLED -> STALL_NEXT
```
If we move the stall check into before we read the fetchq, then we could get the stall_flag bit and shift it into position such that we don't need the branch.  That is, use STALL_NEXT=2, STALLED=1, NOT_STALLED=0.  Shift the flag so that if set we are STALL_NEXT, always shift right to get to STALLED and clear w/ an assignment

kmabeeTT replied:
This makes sense.. updates enums, and using this for handling of stall_flag as per suggestion:

```
    bool stall_flag = (prefetch_q_rd_ptr_local & prefetch_q_msb_mask) != 0;
    stall_state = stall_flag << 1; // NOT_STALLED -> STALL_NEXT if stall_flag is set
```

However, don't see a way ariound this branch on STALL_NEXT handling...

```
        read_from_pcie<preamble_size>
            (prefetch_q_rd_ptr, pending_read_size, fence, pcie_read_ptr, cmd_ptr, fetch_size);
        if (stall_state == STALL_NEXT) {
            barrier_and_stall(pending_read_size, fence); // STALL_NEXT -> STALLED
        }

```

... so might as well use asignment instead of cryptic right shift I figure to reach `STALLED`, inside this function (code moved out of read_from_pcie()) :

```
// If prefetcher must stall after this fetch, wait for data to come back, and move to stalled state.
FORCE_INLINE
void barrier_and_stall(uint32_t& pending_read_size, uint32_t& fence) {
    noc_async_read_barrier();
    fence += pending_read_size;
    pending_read_size = 0;
    stall_state = STALLED;
}
```

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -158,8 +153,13 @@ void read_from_pcie(volatile tt_l1_ptr uint32_t *& prefetch_q_rd_ptr,
 
-    // NOT_STALLED Persists. STALL_NEXT -> STALLED.
-    stall_state = static_cast<StallState>(stall_state << 1);
+    // If prefetcher must stall after this fetch, wait for data to come back, and move to next state.
```
I don't like this version as much. I think the "stall" concept is a "get cmds" concept and "read_from_pcie just does reads

kmabeeTT replied:
Sounds good, was mainly put here to prevent duplication, will move to right after read_from_pcie() call.

kmabeeTT replied:
I see, was mainly to reduce needing to duplicate the same code twice by caller, will stick in inline function and call twice.

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -879,6 +876,11 @@ bool process_cmd(uint32_t& cmd_ptr,
 
+    // Done processing the outer command, stall is no longer required
+    if (!exec_buf) {
+        stall_state = NOT_STALLED;
```
I prefer hanging.  We could use an assert here to enforce the behavior we want, ASSERT(exec_buf || stall_state != NOT_STALLED).  Using this concept incorrectly should error out rather than work as that may hide a real bug.

kmabeeTT replied:
Prefer hangs, got it.  Can move the setting of stall_state = NOT_STALLED back to CQ_PREFETCH_CMD_EXEC_BUF handling no problem.  

As for assert in the end of `process_cmd()`... not finding a clean way to do this. It's almost like this (flipped polarity of what you said, and also considered the ExecBuf outer cmd itself):

```
    // Only ExecBuf cmds itself or the subcmds within ExecBuf cmds can cause Prefetcher Stall.
    ASSERT(cmd->base.cmd_id == CQ_PREFETCH_CMD_EXEC_BUF || exec_buf || stall_state == NOT_STALLED);
```

But not quite that narrow, because currently we bundle a couple commands in the same FetchQ entry with the "stall_flag=1", so this assert fires on handling of `CQ_PREFETCH_CMD_RELAY_INLINE` which is allwed to be in stalled state since it preceeds an ExecBuf cmd:


```
void EnqueueTraceCommand::process() {
    uint32_t cmd_sequence_sizeB =
        CQ_PREFETCH_CMD_BARE_MIN_SIZE + // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WAIT
        CQ_PREFETCH_CMD_BARE_MIN_SIZE;  // CQ_PREFETCH_CMD_EXEC_BUF
```

I didn't include Tony's change to split `ExecBuf` to it's own cmd after we said it wasnt necessary. 

Hmm... skip assert idea, and just let hang?

pgkeller replied:
Oh, right, my bad.  Of course we enter the stall state not in exec buf, that's what it's there for.  Hmm.  Yes, if not easy to assert correctness let it fail, I'd rather there be an indicator that someone is doing something wrong than gloss over it

pgkeller reviewed (commented):
not 100% sure my proposal will work, let me know if not (the avoid "ifs" idea)

kmabeeTT commented on tt_metal/impl/dispatch/kernels/cq_dispatch.cpp:
```diff
@@ -665,7 +665,9 @@ static void process_wait() {
     if (clear_count) {
-        *sem_addr = 0;
+        uint32_t neg_sem_val = -(*sem_addr);
+        noc_semaphore_inc(get_noc_addr_helper(my_noc_xy, get_semaphore(upstream_sync_sem)), neg_sem_val);
```
This looks like not the correct change, and it is responsible for hangs.  I might have not understood it when I added it, and put the wrong semaphore/address in this call.   I think I should just pass `DISPATCH_MESSAGE_ADDR` as the address instead of `get_semaphore(upstream_sync_sem)` ? Trying that.

pgkeller replied:
That looks right to me

kmabeeTT commented:
test_prefetcher smoke test was hanging, probably because EXEC_BUF is not setting stall_flag=1 (MSB in FetchQ entry), so I added assert in cq_prefetcher just now to catch when exec_buf is seen and we are not in STALLED state (we should be) and sure enough it fires.  TODO: Update test_prefetcher.cpp ...somewhere.... to account for this.

kmabeeTT commented:
> > all-post-commit CI run: https://github.com/tenstorrent/tt-metal/actions/runs/8962469388
> > test_prefetcher smoke test was hanging, probably because EXEC_BUF is not setting stall_flag=1 (MSB in FetchQ entry), so I added assert in cq_prefetcher just now to catch when exec_buf is seen and we are not in STALLED state (we should be) and sure enough it fires. TODO: Update test_prefetcher.cpp ...somewhere.... to account for this.
> 
> It's a little ugly, but at the end of gen_prefetcher_exec_buf_cmd, as the last line, you could OR the flag into the previous cmd_sizes entry, I think that'll work

Thanks, yeah I arrived at same conclusion.  Updated test for this (and a tiny bit more as per our offline discussion, needed to change cmd_sizes from uint16_t to uint32_t to carry MSB since FetchQ entry is 32 bit right now) and now test is passing again. Squashed changes, should be release candidate now if you can take another final peak plz.

kmabeeTT commented:
Unfortunately one more issue here:  test_prefetcher smoke test is hanging on ExecBuf when run in "split prefetcher" mode via `-spre` flag, CI link:

https://github.com/tenstorrent/tt-metal/actions/runs/8986425622/job/24682654291#step:8:1372

CMD:

```
./build/test/tt_metal/perf_microbenchmark/dispatch/test_prefetcher -t 1 -i 5 -x -spre
```

With watcher, It hits my assert to show that the "stall flag" is missing from FetchQ entry MSB. My test change is somehow not getting applied properly in split prefetcher mode.  I have no experience with split prefetcher mode yet, so I don't know what fix is (test fix or prefetcher kernel fix).  Got this feedback from @pgkeller :

```
So, split prefetcher is splitting to the _h and _d versions.  The issue is that w/o a stall, prefetch_h could keep sending cmds to prefetch_d while prefetch_d is trying to use the same buffer for exec_buf commands coming from dram.  So, prefetch_h stalls w/ the exec_buf and prefetch_d release it at the end of the trace.  In test_prefetcher, this is handled in gen_prefetcher_exec_buf_cmd_and_write_to_dram (I wish I had more comments there).  Guessing something is going wrong w/ this pathway
```

Trying to look more, open to suggestions.

kmabeeTT committed:
#8062: Fix for test_prefetcher split_prefetcher ExecBuf issues

 - Hang (assert with watcher) that STALL state wasn't seen when
   handling ExecBuf

 - Don't know if this is correct fix, but it seems to work...

kmabeeTT commented:
I pushed a possible fix, change to cq_prefetch.cpp that seems to work (but I don't know why), can you double check if it's the right fix, @pgkeller  ?

kmabeeTT commented:
Hey @pgkeller, wanted to circle back here.. my changes passed [all-post-commit](https://github.com/tenstorrent/tt-metal/actions/runs/8994678611) after I submitted that fix, you mentioned offline it looked good, but that split-prefetcher doesn't need the stall feature.  Should I go ahead and merge, or leave it open and someone can work on making these changes not apply to split prefetcher somehow?  Don't think I'll get a chance to spend anymore time here soon.

FYI @tt-aho

pgkeller commented:
Let's sit on this for a bit - there are some other changes coming for prefetcher that interact and I'm not sure how that will all settle out

tt-aho commented:
These changes were merged with a different pr that needed these changes to fix tests.
